### PR TITLE
Ignoring messages with timestamps in the past or in the future

### PR DESCRIPTION
**Pull request by rwrzesien**
Resolves https://github.com/golemfactory/concent/issues/45
https://www.pivotaltracker.com/story/show/154403513

### CODE REVIEW

File: concent_api/core/views.py [hunk 1]
```diff
@@ -17,6 +21,8 @@
 @api_view
 @require_POST
 def send(_request, client_message):
+    if client_message:
```
Comment (cameel) [inline concent_api/core/views.py]: Client message is not a boolean value.

File: concent_api/core/tests/test_core_views.py [hunk 2]
```diff
@@ -286,6 +286,40 @@ def test_send_reject_message_save_as_receive_out_of_band_status(self):
         self.assertEqual(reject_response.status_code, 202)
         self.assertEqual(Message.objects.last().type, message.RejectReportComputedTask.TYPE)
 
+    @freeze_time("2017-11-17 10:00:00")
+    def test_send_reject_message_when_timestamp_too_old(self):
```
Comment (cameel) [inline concent_api/core/tests/test_core_views.py]: You're missing a `should` here.

File: concent_api/core/tests/test_core_views.py [hunk 3]
```diff
@@ -286,6 +286,40 @@ def test_send_reject_message_save_as_receive_out_of_band_status(self):
...
+        response = self.client.post(
+            reverse('core:send'),
+            data = dump(
+                message.Ping(
+                    timestamp=int(dateutil.parser.parse("2017-11-17 09:40:00").timestamp())
+                ),
+                PROVIDER_PRIVATE_KEY,
+                CONCENT_PUBLIC_KEY),
```
Comment (cameel) [inline concent_api/core/tests/test_core_views.py]: Please fix your style here.

File: concent_api/core/tests/test_core_views.py [hunk 4]
```diff
@@ -286,6 +286,40 @@ def test_send_reject_message_save_as_receive_out_of_band_status(self):
         self.assertEqual(reject_response.status_code, 202)
         self.assertEqual(Message.objects.last().type, message.RejectReportComputedTask.TYPE)
 
+    @freeze_time("2017-11-17 10:00:00")
```
Comment (cameel) [inline concent_api/core/tests/test_core_views.py]: Please use `with freeze_time()` around the tested call instead of wrapping the whole function.

File: concent_api/core/tests/test_core_views.py [hunk 5]
```diff
@@ -286,6 +286,40 @@ def test_send_reject_message_save_as_receive_out_of_band_status(self):
...
+    def test_send_reject_message_when_timestamp_too_old(self):
+        response = self.client.post(
+            reverse('core:send'),
+            data = dump(
+                message.Ping(
+                    timestamp=int(dateutil.parser.parse("2017-11-17 09:40:00").timestamp())
```
Comment (cameel) [inline concent_api/core/tests/test_core_views.py]: You're missing an `assert` here. You should assert that the difference between the current time and your timestamp is greater than `MSG_TTL` defined in golem-messages. Currently it is but if this ever gets changed our tests will silently just stop working.

File: concent_api/core/tests/test_core_views.py [hunk 6]
```diff
@@ -286,6 +286,40 @@ def test_send_reject_message_save_as_receive_out_of_band_status(self):
...
+        self.assertEqual(response.status_code, 400)
+        self.assertIn('error', response.json().keys())
+
+    @freeze_time("2017-11-17 10:00:00")
+    def test_send_reject_message_when_timestamp_too_far_in_future(self):
+        response = self.client.post(
+            reverse('core:send'),
+            data = dump(
+                message.Ping(
+                    timestamp=int(dateutil.parser.parse("2017-11-17 10:10:00").timestamp())
```
Comment (cameel) [inline concent_api/core/tests/test_core_views.py]: You're missing an `assert` here. You should assert that the difference between the current time and your timestamp is greater than `FUTURE_TIME_TOLERANCE` defined in golem-messages. Currently it is but if this ever gets changed our tests will silently just stop working.

File: concent_api/core/tests/test_core_views.py [hunk 7]
```diff
@@ -286,6 +286,40 @@ def test_send_reject_message_save_as_receive_out_of_band_status(self):
...
+                CONCENT_PUBLIC_KEY),
+            content_type                   = 'application/octet-stream',
+            HTTP_CONCENT_CLIENT_PUBLIC_KEY = b64encode(PROVIDER_PUBLIC_KEY).decode('ascii')
+        )
+
+        self.assertEqual(response.status_code, 400)
+        self.assertIn('error', response.json().keys())
+
+    @freeze_time("2017-11-17 10:00:00")
+    def test_send_reject_message_when_timestamp_too_far_in_future(self):
```
Comment (cameel) [inline concent_api/core/tests/test_core_views.py]: See my remarks for the other test. They mostly apply to this one too.

File: concent_api/core/views.py [hunk 8]
```diff
@@ -374,6 +403,13 @@ def validate_golem_message_reject(data):
             raise Http400("Wrong type of deadline field.")
 
 
+def validate_golem_message_timestamp(timestamp):
+    try:
+        verify_time(timestamp)
+    except (MessageFromFutureError, MessageTooOldError, TimestampError):
+        raise Http400('Message timestamp is wrong.')
```
Comment (cameel) [inline concent_api/core/views.py]: Please return the original exception message instead of replacing it with a generic one that does not even say what's wrong with the timestamp.

File: concent_api/core/views.py (hunk 1 above)
Comment (cameel) [inline concent_api/core/views.py]: Also, `client_message` should never be `None` here (please add an `assert` that checks that). If `api_view` gets `None` after decoding a message, it should return an error. But this is not related to this branch. Please report and fix it separately.

File: concent_api/core/tests/test_core_views.py (hunk 4 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_core_views.py]: :+1:

File: concent_api/core/tests/test_core_views.py (hunk 2 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_core_views.py]: :+1:

File: concent_api/core/views.py (hunk 1 above)
Comment (rwrzesien) [reply concent_api/core/views.py]: Yes, lets make a separate issue out of it. Generally I think it should never hit the view and fail directly in api_views.py. I will work on that later.

Comment (rwrzesien) [reply concent_api/core/views.py]: :+1:

File: concent_api/core/views.py (hunk 8 above)
Comment (rwrzesien) [reply concent_api/core/views.py]: :+1:

File: concent_api/core/tests/test_core_views.py (hunk 3 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_core_views.py]: :+1:

File: concent_api/core/tests/test_core_views.py (hunk 6 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_core_views.py]: :+1:

File: concent_api/core/tests/test_core_views.py (hunk 7 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_core_views.py]: :+1:

File: concent_api/core/tests/test_core_views.py (hunk 5 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_core_views.py]: :+1:

**Comment by rwrzesien** [comment]
@cameel Updated.

File: concent_api/core/views.py [hunk 9]
```diff
@@ -374,6 +403,17 @@ def validate_golem_message_reject(data):
...
 
+def validate_golem_message_timestamp(timestamp):
+    try:
+        verify_time(timestamp)
+    except MessageFromFutureError:
+        raise Http400('Message timestamp too much in the future.')
+    except MessageTooOldError:
+        raise Http400('Message timestamp too old.')
+    except TimestampError as e:
+        raise Http400(e)
```
Comment (cameel) [inline concent_api/core/views.py]: Please don't use single-letter names for your variables.

File: concent_api/core/views.py [hunk 10]
```diff
@@ -82,19 +88,24 @@ def send(_request, client_message):
         force_report_computed_task = message.Message.deserialize(
             force_report_computed_task_from_database.last().data.tobytes(),
             None,
-            check_time = False
```
Comment (cameel) [inline concent_api/core/views.py]: Actually, now that I think about it this line is wrong.

You're removing a line that ignores the timestamp on a message deserialized from the database. We **do** want to ignore the timestamp here.

What you should be doing instead is to remove `check_time` in `api_view` decorator which decoded the outer message for the view.

File: concent_api/core/views.py (hunk 9 above)
Comment (rwrzesien) [reply concent_api/core/views.py]: :+1:

**Comment by rwrzesien** [comment]
@cameel Yes, it works, and leads to message being None, for which we have separate issue. Tests I have written for this issue covers this case.
